### PR TITLE
RUMM-2270 Introduce Session Replay Recorder

### PR DESCRIPTION
**Pull request by ncreated**
### What and why?

📦 This PR brings the scaffold 🏗️ of Session Replay architecture. Following the RFC, it implements basics of the `Recorder` and scratches the surface of the `Processor`.

### How?

The foundation I made distinguishes several components:
* `Recorder` - the heart beat 💓 of SR:
   * `ViewTreeSnapshotProducer` - produces view tree _snapshots_,
   * `Scheduler` - runs recurring operation of passing _snapshots_ to `Processor`,
* `Processor: ViewTreeSnapshotProcessor` - it will reduce _snapshots_ into sequence of mutation records send to SR BE.

Both tiers are glued together using `ViewTreeSnapshot` model. Produced in `ViewTreeSnapshotProducer` and received in `Processor`, it stands for an intermediate representation of the native view tree (before it is turned into wireframe definitions from [rum-events-format](https://github.com/DataDog/rum-events-format)):
```
[native views hierarchy] → [ViewTreeSnapshot] → [wireframes]
```
Unlike native views, the `ViewTreeSnapshot` will be developed with thread safety in mind. The upcoming implementation of `Processor` will leverage this fact and use background thread for flattening the view tree information (in my POC this was proven to be the heaviest part of SR capturing and if ran on the main thread, it was visibly impacting the app FPS).

### Review checklist
- [ ] Feature or bugfix MUST have appropriate tests (unit, integration)
- [ ] Make sure each commit and the PR mention the Issue number or JIRA reference
- [ ] Add CHANGELOG entry for user facing changes

### Custom CI job configuration (optional)
- [ ] Run unit tests
- [ ] Run integration tests
- [ ] Run smoke tests

### CODE REVIEW

File: session-replay/Sources/DatadogSessionReplay/Utilities/Schedulers/MainThreadScheduler.swift [hunk 1]
```diff
@@ -0,0 +1,72 @@
...
+/// It uses the main `RunLoop` as its time source, meaning that all operations are suspended
+/// while the app is not in foreground.
+internal class MainThreadScheduler: Scheduler {
+    private struct Constants {
+        /// The `tolerance` applied to `Timer` for reducing the impact on the power usage of the app.
+        ///
+        /// Ref.: https://developer.apple.com/documentation/foundation/timer
+        /// > A general rule, set the tolerance to at least 10% of the interval, for a repeating timer.
+        /// > Even a small amount of tolerance has significant positive impact on the power usage of the application.
+        static let timerTolerance: Double = 0.1
```
Comment (mariusc83) [inline session-replay/Sources/DatadogSessionReplay/Utilities/Schedulers/MainThreadScheduler.swift:20]: Nice

**Review by mariusc83 — APPROVED** [review]
LGTM

File: Makefile [hunk 2]
```diff
@@ -70,6 +70,11 @@ xcodeproj-httpservermock:
 		@cd instrumented-tests/http-server-mock/ && swift package generate-xcodeproj
 		@echo "OK 👌"
 
+xcodeproj-session-replay:
+		@echo "⚙️  Generating 'DatadogSessionReplay.xcodeproj'..."
+		@cd session-replay/ && swift package generate-xcodeproj
```
Comment (maxep) [inline Makefile:75]: This is great! I would like to leverage this further in v2 when we start restructuring the project structure!

**Review by maxep — APPROVED** [review]
Awesome! it's well designed and documented 👌 Looking forward to see it grow!

File: session-replay/Sources/DatadogSessionReplay/Recorder/ViewTree/ViewTreeSnapshotProducer.swift [hunk 3]
```diff
@@ -0,0 +1,12 @@
...
+ * Unless explicitly stated otherwise all files in this repository are licensed under the Apache License Version 2.0.
+ * This product includes software developed at Datadog (https://www.datadoghq.com/).
+ * Copyright 2019-2020 Datadog, Inc.
+ */
+
+import Foundation
+
+/// Produces `ViewTreeSnapshot` describing the user interface.
+internal protocol ViewTreeSnapshotProducer {
+    func takeSnapshot() -> ViewTreeSnapshot?
```
[the remark this answers is not in the document]
Comment (ncreated) [reply session-replay/Sources/DatadogSessionReplay/Recorder/ViewTree/ViewTreeSnapshotProducer.swift:11]: Good point 👍. Whereas I totally agree that all problems related to building `VTS` in our logic should be reported as errors, in this early phase (no `VTS` building yet) the optional originates in `windows.filter { $0.isKeyWindow }` being an optional. I wonder if this is correct to assume that there must be always a key window (e.g. knowing app state transitions or iPadOS split screen). Any thoughts @maxep ?

**Comment by ncreated** [comment]
CI is green, Bitrise Check fails (again 😞) to report the right status.

Screenshot 2022-08-09 at 12 42 44

File: session-replay/Sources/DatadogSessionReplay/Recorder/ViewTree/ViewTreeSnapshotProducer.swift (hunk 3 above)
[the remark this answers is not in the document]
Comment (maxep) [reply session-replay/Sources/DatadogSessionReplay/Recorder/ViewTree/ViewTreeSnapshotProducer.swift:11]: I don't mean assuming there must always be a window, and the notion of window is abstracted away from `ViewTreeSnapshotProducer`. Throwing an error gives the opportunity to producers to report any reason why it has failed to take a snapshot. Right now, it's the `Recorder` that assumes that the missing snapshot is a result of a missing window, but that might not always be true.
It's not a change request for this PR :) but it something to consider.

[the remark this answers is not in the document]
Comment (ncreated) [reply session-replay/Sources/DatadogSessionReplay/Recorder/ViewTree/ViewTreeSnapshotProducer.swift:11]: Got it, very clear 👍. I totally agree - error throwing should be preferred for more clarity. I will see how it goes when going deeper into `VTSProducer` implementation.

**Pull request merged**